templates/typescript_gapic/src/$version/$service_client.ts.njk

```diff
@@ -0,0 +1,303 @@
+    let protoFilesRoot = new gax.GoogleProtoFilesRoot();
+{%- for filename in api.filesToGenerate %}
+    protoFilesRoot = protobuf.loadSync(
+      path.join(__dirname, '..', '..', 'protos', '{{ filename }}'),
+      protoFilesRoot);
+{%- endfor %}
+
+    const operationsClient = new gax.lro({
+      auth: gaxGrpc.auth,
+      grpc: gaxGrpc.grpc,
+    }).operationsClient(opts);
+
+{%- for method in service.longrunning %}
+    const {{ method.name.toCamelCase() }}Response = protoFilesRoot.lookup(
+      '{{ method.longrunning.responseType }}');
+    const {{ method.name.toCamelCase() }}Metadata = protoFilesRoot.lookup(
+      '{{ method.longrunning.metadataType }}');
+{%- endfor %}
+
+    this._descriptors.longrunning = {
+{%- set longrunningJoiner = joiner() %}
+{%- for method in service.longrunning %}
+      {{- longrunningJoiner() }}
+      {{ method.name.toCamelCase() }}: new gax.LongrunningDescriptor(
+        operationsClient,
+        {{ method.name.toCamelCase() }}Response.decode.bind({{ method.name.toCamelCase() }}Response),
+        {{ method.name.toCamelCase() }}Metadata.decode.bind({{ method.name.toCamelCase() }}Metadata))
+{%- endfor %}
+    };
+{%- endif %}
+
+    // Put together the default options sent with requests.
+    const defaults = gaxGrpc.constructSettings(
+        '{{ api.naming.protoPackage }}.{{ service.name }}', gapicConfig as gax.ClientConfig,
+        opts.clientConfig || {}, {'x-goog-api-client': clientHeader.join(' ')});
+
+    // Set up a dictionary of "inner API calls"; the core implementation
+    // of calling the API is handled in `google-gax`, with this code
+    // merely providing the destination and request information.
+    this._innerApiCalls = {};
+
+    // Put together the "service stub" for
+    // google.showcase.v1alpha2.Echo.
+    const {{ service.name.toCamelCase() }}Stub =
+        // @ts-ignore We cannot check types that are loaded in runtime.
+        gaxGrpc.createStub(gaxProtos.{{ api.naming.protoPackage }}.{{ service.name }}, opts);
+
+    const {{ service.name.toCamelCase() }}StubMethods =
+        [
+{%- set stubMethodsJoiner = joiner(', ') -%}
+{%- for method in service.method -%}
+      {{- stubMethodsJoiner() -}}
+      '{{ method.name.toCamelCase() }}'
+{%- endfor -%}
+    ];
+
+    for (const methodName of echoStubMethods) {
+      this._innerApiCalls[methodName] = gax.createApiCall(
+          {{ service.name.toCamelCase() }}Stub.then(
+              // @ts-ignore
+              stub => (...args) => {
+                stub[methodName].apply(stub, args);
+              },
+              (err: Error|null|undefined) => () => {
+                throw err;
+              }),
+          defaults[methodName],
+          this._descriptors.page[methodName] ||
+              this._descriptors.stream[methodName] ||
+              this._descriptors.longrunning[methodName]);
+    }
+  }
+
+  /**
+   * The DNS address for this API service.
+   */
+  static get servicePath() {
```

Comment:
I know this is based off code I wrote, but bear in mind that we need to be able to handle the case where the service does not define a default hostname. See [AIP-4210](https://aip.dev/client-libraries/4210#services-and-methods) for the spec here.

This is really important because there may be a small number of services where the user will be forced to select a (region-based) hostname in the future.